kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -474,23 +479,28 @@ public CompletionStage<ResponseFilterResult> forwardResponse(ResponseHeaderData
             return responseFilterResultBuilder().forward(header, response).completed();
         }
 
+        @NonNull
         @Override
-        public <T extends ApiMessage> CompletionStage<T> sendRequest(short apiVersion, ApiMessage request) {
-            short key = request.apiKey();
-            var apiKey = ApiKeys.forId(key);
-            short headerVersion = apiKey.requestHeaderVersion(apiVersion);
-            var header = new RequestHeaderData()
-                    .setCorrelationId(-1)
-                    .setRequestApiKey(key)
-                    .setRequestApiVersion(apiVersion);
-            if (headerVersion > 1) {
-                header.setClientId(filter.getClass().getSimpleName() + "@" + System.identityHashCode(filter));
-            }
-            boolean hasResponse = apiKey != ApiKeys.PRODUCE
-                    || ((ProduceRequestData) request).acks() != 0;
-            var filterPromise = new InternalCompletableFuture<T>(ctx.executor());
+        public <T extends ApiMessage> CompletionStage<T> sendRequest(short apiVersion, @NonNull ApiMessage request) {
+            return sendRequest(new RequestHeaderData().setRequestApiVersion(apiVersion), request)
+                    .thenApply(r -> r == null ? null : (T) r.message());
+        }
+
+        @NonNull
+        @Override
+        public <M extends ApiMessage> CompletionStage<ResponseHeaderAndApiMessage<M>> sendRequest(@NonNull RequestHeaderData header,
+                                                                                                  @NonNull ApiMessage request) {
+            Objects.requireNonNull(header);
+            Objects.requireNonNull(request);
+
+            var apiKey = ApiKeys.forId(request.apiKey());
+            header.setRequestApiKey(apiKey.id);
+            header.setCorrelationId(-1);
+
```

Comment:
as it stands sendRequest will try to generate a request with an illegal version.

1. fail early if the client knows this is unsupported.
```
            if (!apiKey.isVersionSupported(header.requestApiVersion())) {
                throw ...
            }
```
2. we could go a step further an validate against the api service.

One could argue that server side will validate both conditions, so why bring in the extra complexity (especially for 2).

I'm +1 on the first.
I'm less sure about the wisdom of the second.

Reply:
On the second point, a Filter Author that wants to provide a FIlter implementation that adapts to the runtime environment would need to call `ApiVersionsService` themselves anyway.  They'll need to do this to know how to generate a request object in the right shape and what requestApiVersion to set on the header.  #sendRequest's own call to ApiVersionsService wouldn't benefit.

Reply:
I think we need to be careful about how much validation we perform in the runtime vs how much we let the broker reject. Yes there are a lot of efficiency wins for preventing junk flowing to the broker but I just worry about the overheads of maintaining validation in conjunction with the broker. 

I think this specific change is absolutely fine, but wanted to raise the wider point.